File: ports/go/src/core/graphics/bcsimage.go
package graphics

import (
	"sync"
	core "github.com/robertpelloni/btk-go/ports/go/src/core/kernel"
)

type BcsImage struct {
	*core.BcsObject
	width  int
	height int
	pixels []byte
	mu     sync.RWMutex
}

func NewBcsImage(parent *core.BcsObject, width, height int) *BcsImage {
	return &BcsImage{
		BcsObject: core.NewBcsObject(parent),
		width:     width,
		height:    height,
		pixels:    make([]byte, width*height*4), // Assuming RGBA
	}
}

func (img *BcsImage) Size() (int, int) {
	img.mu.RLock()
	defer img.mu.RUnlock()
	return img.width, img.height
}

func (img *BcsImage) SetPixel(x, y int, r, g, b, a byte) {
	img.mu.Lock()
	defer img.mu.Unlock()
	if x >= 0 && x < img.width && y >= 0 && y < img.height {
		idx := (y*img.width + x) * 4
		img.pixels[idx] = r
		img.pixels[idx+1] = g
		img.pixels[idx+2] = b
		img.pixels[idx+3] = a
	}
}

func (img *BcsImage) Pixels() []byte {
	img.mu.RLock()
	defer img.mu.RUnlock()
	pixelsCopy := make([]byte, len(img.pixels))
	copy(pixelsCopy, img.pixels)
	return pixelsCopy
}

File: ports/rust/src/core/graphics/bcs_image.rs
use std::sync::{Arc, Mutex};
use crate::core::kernel::bcs_object::BcsObject;

pub struct BcsImage {
    pub base: Arc<BcsObject>,
    width: Mutex<i32>,
    height: Mutex<i32>,
    pixels: Mutex<Vec<u8>>,
}

impl BcsImage {
    pub fn new(parent: Option<Arc<BcsObject>>, width: i32, height: i32) -> Arc<Self> {
        let base = if let Some(p) = parent {
            BcsObject::new_with_parent(p)
        } else {
            BcsObject::new()
        };

        Arc::new(Self {
            base,
            width: Mutex::new(width),
            height: Mutex::new(height),
            pixels: Mutex::new(vec![0; (width * height * 4) as usize]), // RGBA
        })
    }

    pub fn size(&self) -> (i32, i32) {
        (*self.width.lock().unwrap(), *self.height.lock().unwrap())
    }

    pub fn set_pixel(&self, x: i32, y: i32, r: u8, g: u8, b: u8, a: u8) {
        let w = *self.width.lock().unwrap();
        let h = *self.height.lock().unwrap();
        if x >= 0 && x < w && y >= 0 && y < h {
            let mut px = self.pixels.lock().unwrap();
            let idx = ((y * w + x) * 4) as usize;
            px[idx] = r;
            px[idx + 1] = g;
            px[idx + 2] = b;
            px[idx + 3] = a;
        }
    }

    pub fn pixels(&self) -> Vec<u8> {
        self.pixels.lock().unwrap().clone()
    }
}
